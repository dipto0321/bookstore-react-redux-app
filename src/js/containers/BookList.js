import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { switchAddBookMode, removeBook, switchProgressUpdate } from '../actions/books';
import { changeFilter } from '../actions/filter';
import Book from '../components/Book';
import CategoryFilter from '../components/CategoryFilter';
import ProgressUpdateModal from '../components/ProgressUpdateModal';

export const BookList = (
  {
    books,
    changeFilter,
    filter,
    removeBook,
    switchAddBookMode,
    progressUpdateMode,
    switchProgressUpdate,
  },
) => (
  <div>
    <ProgressUpdateModal
      progressUpdateMode={progressUpdateMode}
      switchProgressUpdate={switchProgressUpdate}
    />
    <div className="booklist-controls">
      <CategoryFilter
        filter={filter}
        handleChange={optionValue => changeFilter(optionValue)}
      />
      <button
        className="add-book-btn btn-sm"
        type="button"
        onClick={switchAddBookMode}
      >
          Add Book
      </button>
    </div>
    <div>
      {books.map(book => (
        <Book
          key={book.id}
          book={book}
          handleRemove={id => removeBook(id)}
        />
      ))}
    </div>
  </div>
);

const mapStateToProps = state => ({
  books: state.books.filter(book => (state.filter === '' ? true : book.category === state.filter)),
  filter: state.filter,
  progressUpdateMode: state.progressUpdateMode,
});

BookList.propTypes = {
  books: PropTypes.instanceOf(Object).isRequired,
  changeFilter: PropTypes.func.isRequired,
  filter: PropTypes.string.isRequired,
  removeBook: PropTypes.func.isRequired,
  switchAddBookMode: PropTypes.func.isRequired,
  progressUpdateMode: PropTypes.instanceOf(Object).isRequired,
  switchProgressUpdate: PropTypes.func.isRequired,
};

export default connect(
  mapStateToProps,
  {
    removeBook, changeFilter, switchAddBookMode, switchProgressUpdate,
  },
)(BookList);
