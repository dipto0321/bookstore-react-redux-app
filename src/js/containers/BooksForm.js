import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import categories from '../data/bookCategories';

const BooksForm = () => {
  return (
    <form>
      <input type="text"/>
      <select>
        {
          categories.map(category => (
            <option
              key={category}
              value={category}
            >
              { category }
            </option>
          ))
        }
      </select>
      <button 
        type="submit"      
      >
        Add Book
      </button>
    </form>
  )
};

export default BooksForm;
