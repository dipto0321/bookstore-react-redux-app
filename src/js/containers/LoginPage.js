import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { login } from '../thunks/currentUser';

export const LoginPage = ({ login, error, history }) => {
  const [emailOrUsername, setEmailOrUsername] = useState('');
  const [password, setPassword] = useState('');
  const [errorMessage, setErrorMessage] = useState(error);

  useEffect(() => {
    setErrorMessage(error);
  }, [error]);

  const loginUser = (e) => {
    e.preventDefault();
    login({ email_or_username: emailOrUsername, password })
      .then(() => {
        setEmailOrUsername('');
        setPassword('');
        setErrorMessage(error);
        history.push('/');
      }).catch(() => {});
  };

  return (
    <form>
      {
        error && <div className="error">{errorMessage}</div>
      }
      <div>
        <label htmlFor="email">Username/Email:</label>
        <input
          type="email"
          id="email"
          value={emailOrUsername}
          onChange={e => setEmailOrUsername(e.target.value)}
        />
      </div>
      <div>
        <label htmlFor="password">Password:</label>
        <input
          type="password"
          id="password"
          value={password}
          onChange={e => setPassword(e.target.value)}
        />
      </div>
      <button
        type="submit"
        onClick={loginUser}
      >
        Login
      </button>
    </form>
  );
};

const mapStateToProps = state => ({
  error: state.error,
});

export default connect(mapStateToProps, { login })(LoginPage);
