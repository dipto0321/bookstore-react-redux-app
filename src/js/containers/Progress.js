import React from 'react';
import CircularProgressbar from 'react-circular-progressbar';

const Progress = () => {
  const percentage = 66;
  return (
    <div className="progress-body">
      <div className="progress-bar">
        <CircularProgressbar percentage={percentage} />
      </div>
      <div className="progress-percentage">
        <p className="percent-complete">{`${percentage}%`}</p>
        <p className="completed">Completed</p>
      </div>
    </div>
  );
};

export default Progress;
