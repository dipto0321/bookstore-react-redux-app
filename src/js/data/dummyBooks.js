import uuid from 'uuid';

const books = [
  {
    id: uuid(),
    title: 'Rich Dad Poor Dad',
    category: 'Learning',
  },
  {
    id: uuid(),
    title: 'Harry Potter',
    category: 'Action',
  },
];

export default books;
