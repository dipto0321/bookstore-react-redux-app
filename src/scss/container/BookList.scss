.booklist-controls {
  display: flex;
  padding: 2rem;
}

.btn-sm {
  padding: 0;
  max-width: 15%;
}