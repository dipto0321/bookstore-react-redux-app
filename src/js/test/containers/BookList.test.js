import React from 'react';
import { shallow } from 'enzyme';
import { BookList } from '../../containers/BookList';
import dummyBooks from '../fixtures/books';
import bookCategories from '../fixtures/categories';

describe('BookList', () => {
  let wrapper;
  let books;
  let categories;
  let changeFilter;
  let filter;
  let removeBook;

  beforeEach(() => {
    books = dummyBooks;
    categories = bookCategories;
    filter = '';
    changeFilter = jest.fn();
    removeBook = jest.fn();
    wrapper = shallow(
      <BookList
        books={books}
        filter={filter}
        changeFilter={changeFilter}
        removeBook={removeBook}
        switchAddBookMode={false}
      />,
    );
  });

  test('it should render properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('it should call changeFilters with new category', () => {
    wrapper.find('CategoryFilter').prop('handleChange')(categories[0]);
    expect(changeFilter).toHaveBeenLastCalledWith(categories[0]);
  });

  test('it should call removeBook with id of an existing book', () => {
    wrapper.find('Book').at(2).prop('handleRemove')(books[2].id);
    expect(removeBook).toHaveBeenLastCalledWith(books[2].id);
  });
});
