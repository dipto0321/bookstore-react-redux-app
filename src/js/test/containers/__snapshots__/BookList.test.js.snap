// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookList it should render properly 1`] = `
<div>
  <CategoryFilter
    filter=""
    handleChange={[Function]}
  />
  <button
    onClick={[MockFunction]}
    type="button"
  >
    Add Book
  </button>
  <div>
    <Book
      book={
        Object {
          "category": "Action",
          "id": 0,
          "title": "Lord of the Rings",
        }
      }
      handleRemove={[Function]}
      key="0"
    />
    <Book
      book={
        Object {
          "category": "Learning",
          "id": 1,
          "title": "Guide to Investing",
        }
      }
      handleRemove={[Function]}
      key="1"
    />
    <Book
      book={
        Object {
          "category": "Sci-Fi",
          "id": 2,
          "title": "Star Trek",
        }
      }
      handleRemove={[Function]}
      key="2"
    />
  </div>
</div>
`;
